include/mfmg/dealii/lanczos.templates.hpp

```diff
@@ -0,0 +1,494 @@
+  {
+    matrix[i + n * i] = (double)t_maindiag[i];
+    matrix[i + n * (i - 1)] = (double)t_offdiag[i - 1];
+    matrix[i - 1 + n * i] = (double)t_offdiag[i - 1];
+  }
+
+  // LAPACK eigenvalue/vector solve.
+
+  // NOTE: this part can be replaced if desired with some platform-specific
+  // library.
+  // NOTE: for accuracy we are using double here rather than ScalarType;
+  // may not be needed.
+  // ISSUE: LAPACK has other solvers that might be more efficient here.
+
+  // Do all in double, regardless of ScalarType (for now)
+  const lapack_int info = LAPACKE_dgeev(
+      LAPACK_COL_MAJOR, 'N', 'V', n, matrix.data(), n, t_evals_r.data(),
+      t_evals_i.data(), NULL, n, t_evecs.data(), n);
+
+  // Sort into ascending order
+
+  std::vector<int> sort_index(n);
+  for (int i = 0; i < n; ++i)
+  {
+    sort_index[i] = i;
+  }
+
+  std::sort(sort_index.begin(), sort_index.end(),
+            [&](int i, int j) { return t_evals_r[i] < t_evals_r[j]; });
+
+  // Output requested eigenvalues.
+
+  if (_verbosity > 0)
+  {
+    std::cout.width(4);
+    std::cout << "It " << n;
+    std::cout.precision(4);
+    std::cout << "   evals ";
+    for (int i = 0; i < n && i < num_requested; ++i)
+    {
+      std::cout << " " << std::fixed << t_evals_r[sort_index[i]];
+    }
+    if (n < num_requested)
+    {
+      std::cout << "\n";
+    }
+  }
+
+  // Save results.
+  // FIXME: this does not follow details style, it should do a single thing
+  // unconditionally
+  if (n >= num_requested)
+  {
+    evals.resize(num_requested);
+    evecs.resize(n * num_requested);
+
+    for (int i = 0; i < num_requested; ++i)
+    {
+      const int si = sort_index[i];
+      evals[i] = static_cast<ScalarType>(t_evals_r[si]);
+      double sum = 0;
+      for (int j = 0; j < n; ++j)
+      {
+        sum += t_evecs[j + n * si] * t_evecs[j + n * si];
+      }
+      const double norm = sqrt(sum);
+      for (int j = 0; j < n; ++j)
+      {
+        evecs[j + n * i] = static_cast<ScalarType>(t_evecs[j + n * si] / norm);
+      }
+    }
+  }
+
+#if 0
+
+void LAPACK_dgeev( char* jobvl, char* jobvr, lapack_int* n, double* a,
+                   lapack_int* lda, double* wr, double* wi, double* vl,
+                   lapack_int* ldvl, double* vr, lapack_int* ldvr, double* work,
+                   lapack_int* lwork, lapack_int *info );
+
+lapack_int LAPACKE_dgeev( int matrix_order, char jobvl, char jobvr,
+                          lapack_int n, double* a, lapack_int lda, double* wr,
+                          double* wi, double* vl, lapack_int ldvl, double* vr,
+                          lapack_int ldvr );
+
+lapack_int LAPACKE_dgeev_work( int matrix_order, char jobvl, char jobvr,
+                               lapack_int n, double* a, lapack_int lda,
+                               double* wr, double* wi, double* vl,
+                               lapack_int ldvl, double* vr, lapack_int ldvr,
+                               double* work, lapack_int lwork );
```

Comment:
What is the deal with these declarations?  Aren't they provided by the header `<lapacke.h>`?